### PR TITLE
Issue 16161

### LINKED ISSUE #16161: Integral of x*sec(x)**2 doesn't evaluate
This comes from #11131.

This integral doesn't evaluate but Wolfram gives an answer that differentiates back to a form equivalent to the integrand:
```julia
In [14]: f = x*sec(x)**2                                                                                                                                      

In [15]: F = x*tan(x) + log(cos(x))                                                                                                                           

In [16]: f                                                                                                                                                    
Out[16]: 
     2   
x⋅sec (x)

In [17]: F                                                                                                                                                    
Out[17]: x⋅tan(x) + log(cos(x))

In [18]: integrate(f, x)                                                                                                                                      
Out[18]: 
⌠             
⎮      2      
⎮ x⋅sec (x) dx
⌡             

In [19]: F.diff(x)                                                                                                                                            
Out[19]: 
  ⎛   2       ⎞   sin(x)         
x⋅⎝tan (x) + 1⎠ - ────── + tan(x)
                  cos(x)         

In [20]: simplify(F.diff(x)-f)                                                                                                                                
Out[20]: 0
```

### PR DESCRIPTION
**Pull request by Teut2711**
#### Brief description of what is fixed or changed
Addition of integration_by_parts

#### Release Notes

* integrals
   * integration_by_parts

### CODE REVIEW

**Comment by XtremeGood** [comment]
#16161

**Comment by pragyanmehrotra** [comment]
Please fix your release notes, " issue 16161.py" is not a valid header. In your case the valid header would be "integrals". To check the list of valid headers please look [here](
https://github.com/sympy/sympy-bot/blob/master/sympy_bot/submodules.txt)

**Comment by czgdp1807** [comment]
@XtremeGood Please read [this](https://github.com/sympy/sympy/wiki/Introduction-to-contributing).

File: sympy/complexanalysis/Contour Integration.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    def solveit_cir(self):
+
+        eqn_residues = self.eq.as_numer_denom()[1]
+        residues = roots(eqn_residues)
+        sum_residues = 0
+        x = list(self.eq.atoms(Symbol))[0]
+        for i, j in residues.items():   
+            if im(i) > 0:    
+                sum_residues += self.residue_calc(i, j, x)
+        return simplify(2*pi*I*sum_residues)
```
Comment (czgdp1807) [inline sympy/complexanalysis/Contour Integration.py]: Probably a new line is needed here.

File: sympy/complexanalysis/Contour Integration.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+from sympy import *
+import matplotlib.pyplot as plt
+
+class ComplexIntegral:
+
+    def __init__(self, eq, params=dict(shape="semi-circle", origin=(0,0), radius=oo, angle=pi)):
```
Comment (czgdp1807) [inline sympy/complexanalysis/Contour Integration.py]: For maintaining consistency in the codebase, you can use [**kwargs](https://www.digitalocean.com/community/tutorials/how-to-use-args-and-kwargs-in-python-3) for accepting parameters to function/method.

File: issue16161.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        else:
+            return integrate(eq)
+
+    for i,j in factor_list(eq)[1]:
+        if type(i) in [type(sec(x)), type(cos(x)), type(sin(x)), type(tan(x)), 
+                       type(csc(x)), type((cot(x))), type(exp(x))]:        
+            integrates.append((i, j))
+
+        else:    
+#         if type(i) in ["<class 'sympy.core.power.Pow'>",
```
Comment (czgdp1807) [inline issue16161.py]: For maintaining consistency in the codebase, you can use [isinstance](https://www.programiz.com/python-programming/methods/built-in/isinstance) for such checks.

**Review by czgdp1807 — CHANGES_REQUESTED** [review]

**Comment by XtremeGood** [comment]
There no problems with that since I know those small changes are to be
made. First I need approval of the inclusion of by parts.

File: issue16161.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+from sympy import *
```
Comment (czgdp1807) [inline issue16161.py]: Please import only the required functionalities for better efficiency of the code.

File: sympy/complexanalysis/Contour Integration.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
+from sympy import *
+import matplotlib.pyplot as plt
+
+class ComplexIntegral:
```
Comment (czgdp1807) [inline sympy/complexanalysis/Contour Integration.py]: Probably inheriting `Basic` would be better.

File: sympy/complexanalysis/Contour Integration.py (hunk 2 above)
Comment (XtremeGood) [reply sympy/complexanalysis/Contour Integration.py]: Actually this is different code and on a different branch. How do I avoid it from being visible here?

File: issue16161.py (hunk 3 above)
Comment (XtremeGood) [reply issue16161.py]: Ok.

File: issue16161.py (hunk 4 above)
Comment (XtremeGood) [reply issue16161.py]: Yes, I know this. I have seen the module.

File: sympy/complexanalysis/Contour Integration.py (hunk 5 above)
Comment (XtremeGood) [reply sympy/complexanalysis/Contour Integration.py]: These files are from another pull request for a different issue. I don't know why these are showing up here. This PR is only meant to be for by_parts_integration, not contour integration.

Comment (czgdp1807) [reply sympy/complexanalysis/Contour Integration.py]: Maybe you are working from the same branch on your system for both the PRs.

Comment (XtremeGood) [reply sympy/complexanalysis/Contour Integration.py]: This file is not in my cloned repo. I have made this on the desktop and added it to a new branch from there not through bash but github.com.

Comment (czgdp1807) [reply sympy/complexanalysis/Contour Integration.py]: Can't say, most probably, it is something related to branch issues.

Comment (XtremeGood) [reply sympy/complexanalysis/Contour Integration.py]: I think now it's gone.

**Comment by XtremeGood** [comment]
That's strange the code isn't ready yet.Can someone tell me what's happening?

**Comment by XtremeGood** [comment]
Why is  https://codecov.io/gh/sympy/sympy/pull/16379/changes showing me changes in modules like physics, combinatorics etc which I haven't even touched?

**Comment by asmeurer** [comment]
Integration by parts is already implemented in the manualintegrate module. If there is an integral that can be solved by parts that integrate cannot compute, then the manualintegrate code should be improved.

> Why is https://codecov.io/gh/sympy/sympy/pull/16379/changes showing me changes in modules like physics, combinatorics etc which I haven't even touched?

https://github.com/sympy/sympy/issues/16196

**Comment by XtremeGood** [comment]
Ok. There is only more method known as Feynman's trick(basically replace x with ax and try to differentiate wrt a, try integrating wrt to x and then with ). Should it be implemented?
https://github.com/sympy/sympy/issues/16084

**Pull request closed without merging**